backend/operations_scanner.go

```diff
@@ -0,0 +1,346 @@
+
+	for _, doc := range s.activeOperations {
+		select {
+		case <-stop:
+			break
+		default:
+			var requeue bool
+			var err error
+
+			opLogger := logger.With(
+				"operation", doc.Request,
+				"operation_id", doc.ID,
+				"resource_id", doc.ExternalID.String(),
+				"internal_id", doc.InternalID.String())
+
+			switch doc.InternalID.Kind() {
+			case cmv1.ClusterKind:
+				requeue, err = s.pollClusterOperation(ctx, opLogger, doc)
+			case cmv1.NodePoolKind:
+				requeue, err = s.pollNodePoolOperation(ctx, opLogger, doc)
+			}
+			if requeue {
+				activeOperations = append(activeOperations, doc)
+			}
+			if err != nil {
+				opLogger.Error(fmt.Sprintf("Error while polling operation '%s': %s", doc.ID, err.Error()))
+			}
+		}
+	}
+
+	s.activeOperations = activeOperations
+}
+
+func (s *OperationsScanner) pollClusterOperation(ctx context.Context, logger *slog.Logger, doc *database.OperationDocument) (requeue bool, err error) {
+	requeue = true
+
+	clusterStatus, err := s.clusterService.GetCSClusterStatus(ctx, doc.InternalID)
+	if err != nil {
+		var ocmError *ocmerrors.Error
+		if errors.As(err, &ocmError) && ocmError.Status() == http.StatusNotFound && doc.Request == database.OperationRequestDelete {
+			err = s.withSubscriptionLock(ctx, logger, doc.OperationID.SubscriptionID, func(ctx context.Context) error {
+				return s.deleteOperationCompleted(ctx, logger, doc)
+			})
+			if err == nil {
+				requeue = false
+			}
+		}
+		return
+	}
+
+	var opStatus arm.ProvisioningState = doc.Status
+	var opError *arm.CloudErrorBody
+
+	switch clusterStatus.State() {
+	case cmv1.ClusterStateError:
+		opStatus = arm.ProvisioningStateFailed
+		// FIXME This is guesswork. Need clarity from Cluster Service
+		//       on what provision error codes are possible so we can
+		//       translate to an appropriate cloud error code.
+		code := clusterStatus.ProvisionErrorCode()
+		if code == "" {
+			code = arm.CloudErrorCodeInternalServerError
+		}
+		message := clusterStatus.ProvisionErrorMessage()
+		if message == "" {
+			message = clusterStatus.Description()
+		}
+		opError = &arm.CloudErrorBody{Code: code, Message: message}
+	case cmv1.ClusterStateInstalling:
+		opStatus = arm.ProvisioningStateProvisioning
+	case cmv1.ClusterStateReady:
+		opStatus = arm.ProvisioningStateSucceeded
+	case cmv1.ClusterStateUninstalling:
+		opStatus = arm.ProvisioningStateDeleting
+	case cmv1.ClusterStatePending, cmv1.ClusterStateValidating:
+		// These are valid cluster states for ARO-HCP but there are
+		// no unique ProvisioningState values for them. They should
+		// only occur when ProvisioningState is Accepted.
+		if opStatus != arm.ProvisioningStateAccepted {
+			logger.Warn(fmt.Sprintf("Got ClusterState '%s' while ProvisioningState was '%s' instead of '%s'", clusterStatus.State(), opStatus, arm.ProvisioningStateAccepted))
+		}
+		return
+	default:
+		logger.Warn(fmt.Sprintf("Unhandled ClusterState '%s' for operation '%s'", clusterStatus.State(), doc.ID))
```

Comment:
You're using `slog` for logging, so please add structured data to the logger appropriately:

```go
logger.With("clusterState", clusterStatus.State(), "operation", doc.ID").Warn("unhandled cluster state")
```

Reply:
(global comment for the whole PR)

Reply:
Also, note that while the default should certainly be to add structured data to a log line, if there's something you _need_ string-formatting for, `log/slog` already will do string formatting without needing to use `fmt.Sprintf`

Reply:
Structured data is added in the calling function using `logger.With`.  That's why I'm passing the `slog.Logger` instance as a function argument instead of making it a struct member.

> `log/slog` already will do string formatting without needing to use `fmt.Sprintf`

No, it won't.  Arguments past the first string argument are key/value pairs.

Reply:
@mbarnes I think my comment was that `doc.ID` seems like exactly the sort of data you'd want to pull out into a structured logger attribute and filter on post-hoc when debugging.

Reply:
(If it's already present in the logger, omit it from string-formatting in the log line, as it's redundant, right?)